test/bash/bash.go

```diff
@@ -1,23 +1,47 @@
 package bash
 
-import "fmt"
+import (
+	"fmt"
+	"strings"
+)
 
-func DiskUsageWithTotal(directory string) string {
-	return fmt.Sprintf("du -c %s", directory)
+type Command []string
+
+func DiskUsageWithTotal(directory string) Command {
+	return Command{"du", "-c", directory}
+}
+
+func FilterLastLineOnly() []string {
+	return Command{"tail", "-n", "1"}
+}
+
+func Pipe(source []string, sink []string) Command {
+	piped := source
+	piped = append(piped, "|")
+	piped = append(piped, sink...)
+	return piped
+}
+
+func ReadFile(path string) Command {
+	return Command{"cat", path}
 }
 
-func FilterLastLineOnly() string {
-	return "tail -n 1"
+func ListDirectory(path string) Command {
+	return Command{"ls", path}
 }
 
-func Pipe(source string, sink string) string {
-	return fmt.Sprintf("%s | %s", source, sink)
+func Shell(command Command) Command {
+	return Command{"sh", "-c", command.String()}
 }
 
-func ReadFile(path string) string {
-	return fmt.Sprintf("cat %s", path)
+func (c Command) String() string {
+	cmdString := ""
+	for _, str := range c {
+		cmdString = fmt.Sprintf("%s %s", cmdString, str)
+	}
+	return strings.Trim(cmdString, " ")
 }
```

Comment:
I think it gives same result
```suggestion
func (c Command) String() string {
	return strings.Join(c, " ")
}
```